rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,710 @@
+
+   - `-Wall`
+   - `-Wno-`
+   - `-Werror=`
+
+   <details><summary>Arguments</summary>
+
+   - (+) Disallowing reuse avoids dealing with change of meaning across versions
+   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
+   - (-) Maintenance burden for everyone using these old constructs, or evaluating old revisions of Nixpkgs.
+   
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7.0
+```
+
+```console
+nix --supported-language-versions
+6.1
+6.2
+6.3
+7.0
+```
+
+## Version interoperability
+
+### Versioned built-ins can be passed across file boundaries
+
+```nix
+# a.nix
+version 6.1;
+builtins.langVersion
+```
+
+```nix
+# b.nix
+version 7.0;
+[ (import ./a.nix) builtins.null ]
+```
+
+```console
+$ nix-instantiate --eval b.nix
+[ 6 null ]
+```
+
+### Expressions are not forward compatible
+
+```nix
+# a.nix
+version 6.1;
+import ./b.nix
+```
+
+```nix
+# b.nix
+version 7.0;
+builtins.null
+```
+
+```console
+$ nix-instantiate --eval a.nix
+error: unsupported Nix language version 7.0
+```
+
+
+### Best-effort interoperability
+
+```nix
+# a.nix
+version 6.1;
+{ increment }:
+increment 1
+```
+
+```nix
+# b.nix
+version 7.0;
+import ./a.nix { increment = x: x + 1 }
+```
+
+Since `increment` written in version 7.0 carries its own implementation with it, forcing it within an expression written in version 6.1 just works:
+
+```console
+$ nix-instantiate --eval b.nix
+2
+```
+
+### Pathological example
+
+Usually existing code will be interacted with by calling functions.
+When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.
+
+```nix
+# a.nix
+version 6.1;
+{ value }:
+value + 1
+```
+
+Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
+
+```nix
+# b.nix
+version 7.0;
+import ./a.nix { value = %5 + 7i%; }
+```
+
+```console
+$ nix-instantiate --eval bnix
+error: unsupported value type `complex` at built-in operator `+`
+```
+
+In the following example, assume version 7.0 *removed* floating point numbers, such that they can no longer be used.
```

Comment:
That situation is covered by "major version bump".

I really don't think we have a persuasive example of the need for this "best effort interoperability" (in addition to which I object to the name -- it's squishy and vague, which is something Nix certainly *isn't*.)

Reply:
Regarding *adding* complex numbers: If Nix were to add complex numbers, surely it would also extend the definition of addition to cover complex-float addition.  As long as old-version code isn't able to *create* complex numbers, the behavior of a 100%-old-code codebase can't change.  So there's no harm in evaluating old code with the newer version of `+` that understands complex numbers.

Reply:
Lastly and least-importantly, the name "best effort interoperability" is squishy and vague, which is something Nix certainly *isn't*.

Reply:
> removing complex numbers should be covered by "major version bump"

This sentence makes no sense to me in this context: The example cited here already contains a major version bump.

> As long as old-version code isn't able to create complex numbers, the behavior of a 100%-old-code codebase can't change. So there's no harm in evaluating old code with the newer version of + that understands complex numbers.

And as long as old code cannot be called with new values like complex numbers.

Reply:
> And as long as old code cannot be called with new values like complex numbers.

Maybe you intend to limit this "best-effort" behavior to "newly-added variants of the [enum `ValueType`](https://github.com/NixOS/nix/blob/8350f06d6c319b250d006b195415b9b6e95d8ca1/src/libexpr/value.hh#L40-L57)".  If so you should say so.  "Best effort" is very open-ended and vague.

Reply:
I'd like to suggest replacing this whole section with a single sentence saying something approximately like: future RFCs which add a language feature must explain how that language feature should interact with code whose `version` declaration is too old to be aware of that feature.

I think that would be much simpler and still ensure that the problems, if any, get solved.